[pytorch][mobile] add NO_EXPORT macro to unset __visibility__ attribute

ljk53 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #25896 [pytorch] update build_android.sh to not build host protoc for libtorch
* #25894 [pytorch] remove pthreadpool dependency in aten/CMake
* #25817 [pytorch][mobile] disable backward function codegen if disable_autograd is set
* **#25816 [pytorch][mobile] add NO_EXPORT macro to unset __visibility__ attribute**
* #25697 [pytorch] introduce INTERN_DISABLE_AUTOGRAD flag to create inference only library for mobile
* #25815 [pytorch][mobile] gate static aten registerer with USE_STATIC_DISPATCH

Summary:
On Android we will release a small set of native APIs designed for mobile use
cases. All of needed libtorch c++ APIs are called from inside this JNI bridge:
android/pytorch_android/src/main/cpp/pytorch_jni.cpp

With NO_EXPORT set for android static library build, it will hide all
original TORCH, CAFFE2, TH/ATen APIs, which will allow linker to strip
out unused ones from mobile library when producing DSO.

If people choose to directly build libtorch DSO then it will still keep
all c++ APIs as the mobile API layer is not part of libtorch build (yet).

Test Plan:
- build libtorch statically and link into demo app;
- confirm that linker can strip out unused APIs;

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25816

Differential Revision: [D17247237](https://our.internmc.facebook.com/intern/diff/D17247237)

ljk53 committed:
[pytorch][mobile] add NO_EXPORT macro to unset __visibility__ attribute

Summary:
On Android we will release a small set of native APIs designed for mobile use
cases. All of needed libtorch c++ APIs are called from inside this JNI bridge:
android/pytorch_android/src/main/cpp/pytorch_jni.cpp

With NO_EXPORT set for android static library build, it will hide all
original TORCH, CAFFE2, TH/ATen APIs, which will allow linker to strip
out unused ones from mobile library when producing DSO.

If people choose to directly build libtorch DSO then it will still keep
all c++ APIs as the mobile API layer is not part of libtorch build (yet).

Test Plan:
- build libtorch statically and link into demo app;
- confirm that linker can strip out unused APIs;

ezyang reviewed (approved):
I'm not sure this is exhaustive, but it's fine as a start.

Will we add some sort of test that will audit the number of visible symbols, to make sure they are what we expect?

ljk53 commented:
> I'm not sure this is exhaustive, but it's fine as a start.

Looks like torch/* all use TORCH_API, caffe2/* all use CAFFE2_API - both are now defined as C10_API / C10_EXPORT. There are a few cases directly exported using C10_EXPORT. So redefining C10_EXPORT should cover most of them.

> Will we add some sort of test that will audit the number of visible symbols, to make sure they are what we expect?

I was thinking of tracking the android gradle build size directly. Will see how hard it is to count visible symbols.

ljk53 committed:
Update on "[pytorch][mobile] add NO_EXPORT macro to unset __visibility__ attribute"


Summary:
On Android we will release a small set of native APIs designed for mobile use
cases. All of needed libtorch c++ APIs are called from inside this JNI bridge:
android/pytorch_android/src/main/cpp/pytorch_jni.cpp

With NO_EXPORT set for android static library build, it will hide all
original TORCH, CAFFE2, TH/ATen APIs, which will allow linker to strip
out unused ones from mobile library when producing DSO.

If people choose to directly build libtorch DSO then it will still keep
all c++ APIs as the mobile API layer is not part of libtorch build (yet).

Test Plan:
- build libtorch statically and link into demo app;
- confirm that linker can strip out unused APIs;

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25816

Differential Revision: [D17247237](https://our.internmc.facebook.com/intern/diff/D17247237)

ljk53 committed:
Update on "[pytorch][mobile] add NO_EXPORT macro to unset __visibility__ attribute"


Summary:
On Android we will release a small set of native APIs designed for mobile use
cases. All of needed libtorch c++ APIs are called from inside this JNI bridge:
android/pytorch_android/src/main/cpp/pytorch_jni.cpp

With NO_EXPORT set for android static library build, it will hide all
original TORCH, CAFFE2, TH/ATen APIs, which will allow linker to strip
out unused ones from mobile library when producing DSO.

If people choose to directly build libtorch DSO then it will still keep
all c++ APIs as the mobile API layer is not part of libtorch build (yet).

Test Plan:
- build libtorch statically and link into demo app;
- confirm that linker can strip out unused APIs;

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25816

Differential Revision: [D17247237](https://our.internmc.facebook.com/intern/diff/D17247237)

ljk53 committed:
Update on "[pytorch][mobile] add NO_EXPORT macro to unset __visibility__ attribute"


Summary:
On Android we will release a small set of native APIs designed for mobile use
cases. All of needed libtorch c++ APIs are called from inside this JNI bridge:
android/pytorch_android/src/main/cpp/pytorch_jni.cpp

With NO_EXPORT set for android static library build, it will hide all
original TORCH, CAFFE2, TH/ATen APIs, which will allow linker to strip
out unused ones from mobile library when producing DSO.

If people choose to directly build libtorch DSO then it will still keep
all c++ APIs as the mobile API layer is not part of libtorch build (yet).

Test Plan:
- build libtorch statically and link into demo app;
- confirm that linker can strip out unused APIs;

Pull Request resolved: https://github.com/pytorch/pytorch/pull/25816

Differential Revision: [D17247237](https://our.internmc.facebook.com/intern/diff/D17247237)

ezyang commented:
> I was thinking of tracking the android gradle build size directly.

That sounds good. @mingbowan has recently put our binary sizes in scuba, and might know what we should do here.
